### DISCUSSION TITLE
Teensy4.1 6-axis instead of 5 steppers and spindle?

### DISCUSSION THREAD

**Discussion by hamyyy** [terjeio/grblHAL#152] [Q&A] [answered]
Is there an easy way to setup grblHAL (on the teensy 4.1) to have 6-axis by replacing the spindle pins with En Stp and Dir pins for a 6th stepper? I'm not planning on making a conventional cnc machine like a mill or laser cutter. some axis are rotational too, so I don't need all the limit switch pins (starting position doesn't matter for those axes)

Is it a simple case of 
- changing the pin mapping 
- setting `#define N_AXIS 6`
- adding `pinModeOutput()` to driver.c

then it should magically work? is there something I'm missing? or is this just not possible with a T4.1?

@phil-barrett, I'm using your Teensy4.1 breakout board for grblHAL. its been great so far. I'm just wondering if the "spindle" headers are wired in the same way as the 5 stepper motor headers (i.e. do they also have safety measures like optocouplers etc). If this mod does work, will It be fine to use the spindle header for another stepper motor? Thanks

**Comment by terjeio** [comment, marked as answer]
> then it should magically work?

Not without a bit more changes to driver.c as it currently does not have any code for c-axis support.
With that in place a modified pin-mapping file should work. The spindle outputs could be rerouted to the aux pins if not in use.

**Reply by hamyyy** [reply]
okay sweet, thanks. I think having a teensy running a 6-axis version of grblHAL will be very useful for things like robotic arms and such. would be useful to make a fork or a PR with 6-axis functionality.

**Reply by terjeio** [reply]
driver.c in the [test branch](https://github.com/terjeio/grblHAL/tree/test) now has C-axis support (untested), you have to modify/create a map file to make it work.

**Reply by hamyyy** [reply]
I've looked through the `driver.c` file. haven't tested it yet. I will test it today. but I noticed that there is no call to `pinModeOutput()` for axis C, only a call to `DIGITAL_OUT()`.

**Reply by terjeio** [reply]
Good catch - fix committed.

**Reply by hamyyy** [reply]
another thing to point out is, the spindle pwm pin can only be used on pins 12 or 13. I'm guessing this is because the timers used are hardcoded. would it be easy to make it so that I can move the pwm pin to one of the aux pins? or would that require too much change in the driver code?

**Reply by phil-barrett** [reply]
Why not move one of the Axis C pins to an Aux pin?

**Reply by hamyyy** [reply]
@phil-barrett I didn't want to do that simply for wiring convenience. I don't have screw terminals yet (they're on their way) so for now, I'm using the other pin headers provided by the breakout board. 

however, I've since decided to against remapping the spindle. Instead, I'll use the aux INPUT pins. I noticed that when using a quadrature encoder, the encoder uses some of the input aux pins, meaning there is already code in place to disable AUX_Input 1 2 and 3. perfect for step dir and EN pins. I just need to make sure the pins are not set as inputs anywhere else in the code.

@terjeio I've tested the `test` branch with 5 axis (to make sure everything still works, which it does) then I modified the mappings for the T4.1. I added three pins and mapped them to pins 30, 34, and 35. I've changed nothing else except:

- bypassing the errors that prevent using 6+ axis with a tennsy4.1 in `driver.h` and `driver.c` files
- setting `N_AXIS` to 6 in `grbl/config.h`
- inverting all enable pins (including C_AXIS_BIT) in `config.h`
- dissabled AUXINPUT_PIN 1 2 and 3 (added `&& (N_AXIS < 6)` in mapping file, see below)
```C
// Define auxillary input pins
#define AUXINPUT0_PIN       (36u) // ST0
#if (!QEI_ENABLE) && (N_AXIS < 6)
#define AUXINPUT1_PIN       (30u) // ST1
#define AUXINPUT2_PIN       (34u) // ST2
#define AUXINPUT3_PIN       (35u) // ST3
#define AUX_N_IN 4
#define AUX_IN_MASK 0b1111
#else
#define AUX_N_IN 1
#define AUX_IN_MASK 0b1
#endif
```

the code compiles and uploads fine. however, the serial connection refuses to work. i cant get IO_sender to connect to it. it just crashes. so does any other serial communication program (Arduino IDE Serial monitor). help?

let me know if you need more information.


Edit: Does IO_sender support 6 axis?

**Reply by hamyyy** [reply]
upon closer inspection, it seems the input AUX pins are protected and seem to have transistors in the circuit. I'm assuming that I cant just set it from input to output without damaging the transistors? @phil-barrett
is there a way I can bypass the transistors easily?

the other (and easier) option is to use AUX 0 1 and 2 (outputs) as @phil-barrett suggested. however, the code for controlling the digital outputs needs to be disabled also to prevent accidental toggles of the pins. (unless i go the hacky way and not worry about the code and simply never trigger the aux outputs manually, would this work @terjeio?)

**Reply by phil-barrett** [reply]
I meant the relay outputs, teensy pins 31, 32 and 33. No way for them to be toggled in default distribution.

**Reply by hamyyy** [reply]
Update!

ive narrowed down the problem to this line of code

```C
int grbl_enter(void)
{
    usb_serialInit();

#ifdef N_TOOLS
    assert(NVS_ADDR_GLOBAL + sizeof(settings_t) + NVS_CRC_BYTES < NVS_ADDR_TOOL_TABLE);
#else
    assert(NVS_ADDR_GLOBAL + sizeof(settings_t) + NVS_CRC_BYTES < NVS_ADDR_PARAMETERS);
#endif
    /* THIS LINE OF CODE >>> */ assert(NVS_ADDR_PARAMETERS + N_CoordinateSystems * (sizeof(coord_data_t) + NVS_CRC_BYTES) < NVS_ADDR_STARTUP_BLOCK);
    assert(NVS_ADDR_STARTUP_BLOCK + N_STARTUP_LINE * (sizeof(stored_line_t) + NVS_CRC_BYTES) < NVS_ADDR_BUILD_INFO);
```

in `grbllib.c`

no idea what to do with this information. commenting out the line seems to make the teensy run normally. but I'm not sure if there are things that don't work now. ill need to do some further testing to find out. IO_sender still does not want to talk to the teensy when running 6-axis

**Reply by terjeio** [reply]
Fix committed, stored line length for $N was too long when 6 axes configured. Sideeffect from G92 offset now beeing stored in NVS?

There is a regression in ioSender as well, try the patched version found [here](http://www.io-engineering.com/downloads/).

> another thing to point out is, the spindle pwm pin can only be used on pins 12 or 13. I'm guessing this is because the timers used are hardcoded. would it be easy to make it so that I can move the pwm pin to one of the aux pins? or would that require too much change in the driver code?

Yes, timers/pins are hardcoded for PWM. It should be possible to add further options using flexPWM for some of the aux pins. I do not think there will be too much change in the driver code for that. However, adding the needed code may take a bit of time to get right.

**Comment by phil-barrett** [comment]
Yes,  SPWM, SPDI and SPEN could be used for STPC, DIRC and STENC.  These are 5V outputs. You will see the Teensy LED for what ever signal you attach to SPWM.

**Comment by hamyyy** [comment]
hi @terjeio,

I've managed to finally spend some time on my project. i tried the patched version of ioSender that you provided in a previous comment, but still no luck. this time ioSender recognises the serial port and opens. the X Y and Z axis shown on the left. but all inputs are blank and the app is unresponsive. after a few seconds, the app closes on its own. any idea on what's happening? thanks.

Edit: sometimes the app doesn't open at all. it stays on a white screen for a few seconds. unresponsive. then closes

**Reply by phil-barrett** [reply]
That is due to the board not responding.  Try using a terminal app like PuTTY or the Arduino Console and see if it responds to $$ or $i.

**Reply by hamyyy** [reply]
I don't think there's an issue with the board. it responds fine when I use a raw serial monitor. I was even able to move all 6 axes using jog commands.

**Reply by terjeio** [reply]
I'll check this tomorrow, it has been a lot of messages to respond to lately...

**Reply by terjeio** [reply]
Can you try the latest [edge version](http://io-engineering.com/downloads) - it runs fine for me with 6 axes enabled in the controller. If this fails too what is the output from the first real-time report after a cold start? Send a `?` to get it. `$I` output may help too.

**Reply by hamyyy** [reply]
the edge version seems to work. I haven't used it beyond just opening it and looking at the console output.
on the main tab, on the left axis A B and C fields are empty. and I cant modify the values by entering them. the "zero" buttons don't seem to work reliably either. but that's a different issue. 

All in all, grblHAL seems to work with 6 axis on a teensy now with integration with ioSender too! pretty cool. I appreciate all the time and effort you guys put into these open source projects.

are there any other g-code senders that are working on compatibility with grblHAL that you know of? would be nice to also try them out to gain ideas as I'm also working on making a specialized sender for my project. thanks :)

**Reply by terjeio** [reply]
>  on the left axis A B and C are empty

Yes, that is a minor bug I have to fix. As soon as there is motion values will be shown. The "Jog" flyout on the right side can be used for jogging all axes.

> I cant modify the values by entering them

Press the \<Enter\> button to set a value, tab or click somewhere outside the field to restore the original value . After a value is shown "zero" works reliably for me.

> are there any other g-code senders that are working on compatibility with grblHAL that you know of?

[bCNC](https://github.com/vlachoudis/bCNC) and [UGS](https://github.com/winder/Universal-G-Code-Sender)

**Reply by terjeio** [reply]
@Hamoudy99 I have just uploaded a new[ edge version](http://io-engineering.com/downloads) with fix for missing values in the DRO.
